spam/spam_tutorial.py

```diff
@@ -140,234 +140,400 @@
+# * **Incorrect**: The number of data points this LF labels incorrectly (if gold labels are provided)
+# * **Empirical Accuracy**: The empirical accuracy of this LF (if gold labels are provided)
+#
+# For _Correct_, _Incorrect_, and _Empirical Accuracy_, we don't want to penalize the LF for examples where it abstained.
+# We calculate these statistics only over those examples where the LF output a label.
 
 # %%
-import numpy as np
+from snorkel.labeling.analysis import LFAnalysis
 
-L_dev = applier.apply(df_dev)
-L_dev_array = L_dev.squeeze()
+LFAnalysis(L=L_train, lfs=lfs).lf_summary()
 
-correct = L_dev_array == Y_dev
-labeled = L_dev_array != ABSTAIN
-accuracy = (correct * labeled).sum() / labeled.sum()
-print(f"Accuracy: {accuracy * 100:.1f}%")
+# %%
+LFAnalysis(L=L_dev, lfs=lfs).lf_summary(Y=Y_dev)
 
 # %% [markdown]
-# Alternatively, you can use the provided `metric_score()` helper method, which allows you to specify a metric to calculate and certain classes to ignore (such as ABSTAIN).
+# So even these very simple rules do quite well!
+# Following our principle above, we might want to pick the `check` rule, since both have high precision and `check` has higher coverage.
+# But let's look at our data to be sure.
+#
+# The helper method `error_buckets()` groups examples by their predicted label and true label. For example, `buckets[(SPAM, HAM)]` contains the indices of data points that the LF labeled `SPAM` that actually belong to class `HAM`. This may give ideas for where the LF could be made more specific.
 
 # %%
-from snorkel.analysis.metrics import metric_score
+from snorkel.analysis.error_analysis import error_buckets
 
-# Calculate accuracy, ignore all examples for which the predicted label is ABSTAIN
-accuracy = metric_score(
-    golds=Y_dev, preds=L_dev_array, metric="accuracy", filter_dict={"preds": [ABSTAIN]}
-)
-print(f"Accuracy: {accuracy * 100:.1f} %")
+buckets = error_buckets(golds=Y_dev, preds=L_dev[:, 1])
+df_dev.iloc[buckets[(SPAM, HAM)]]
 
 # %% [markdown]
-# You can also use the helper class `LFAnalysis()` to report the following summary statistics for multiple LFs at once:
-# * **Polarity**: The set of labels this LF outputs
-# * **Coverage**: The fraction of the dataset the LF labels
-# * **Overlaps**: The fraction of the dataset where this LF and at least one other LF label
-# * **Conflicts**: The fraction of the dataset where this LF and at least one other LF label and disagree
-# * **Correct**: The number of data points this LF labels correctly (if gold labels are provided)
-# * **Incorrect**: The number of data points this LF labels incorrectly (if gold labels are provided)
-# * **Emp. Acc.**: The empirical accuracy of this LF (if gold labels are provided)
-#
-# Since we only have one LF, overlaps and conflicts will be 0.
+# `check` is still looking good, since the false positive is specific to only a few of the most popular videos on YouTube (ones with billions of views).
```

Comment:
But all five our our videos are very popular ones on YouTube. I don't think there was any previous discussion about whether this dataset is representative of YouTube comments at large. I wouldn't bring up that angle; might just confuse people.